File: shared/sessionToken.js
const crypto = require("crypto");

function base64UrlEncode(value) {
  return Buffer.from(value).toString("base64url");
}

function base64UrlDecode(value) {
  return Buffer.from(value, "base64url").toString("utf8");
}

function signValue(value, secret) {
  return crypto.createHmac("sha256", secret).update(value).digest("base64url");
}

/**
 * The session token is intentionally simple and framework-free so both the
 * REST API and realtime service can verify the same payload without adding a
 * JWT dependency. The payload is still signed and expiry-bound.
 */
function signSessionToken(payload, secret) {
  const encodedPayload = base64UrlEncode(JSON.stringify(payload));
  const signature = signValue(encodedPayload, secret);
  return `${encodedPayload}.${signature}`;
}

function verifySessionToken(token, secret) {
  if (!token || typeof token !== "string" || !token.includes(".")) {
    throw new Error("invalid session token");
  }

  const [encodedPayload, signature] = token.split(".");
  const expectedSignature = signValue(encodedPayload, secret);
  const left = Buffer.from(signature);
  const right = Buffer.from(expectedSignature);
  if (left.length !== right.length || !crypto.timingSafeEqual(left, right)) {
    throw new Error("invalid session token signature");
  }

  const payload = JSON.parse(base64UrlDecode(encodedPayload));
  if (typeof payload.exp !== "number" || Date.now() > payload.exp) {
    throw new Error("session token expired");
  }

  return payload;
}

module.exports = {
  signSessionToken,
  verifySessionToken,
};

File: backend_fastapi/app/session_token.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
from typing import Any, Dict


def base64url_encode(value: str) -> str:
    return base64.urlsafe_b64encode(value.encode("utf-8")).decode("utf-8").rstrip("=")


def sign_value(value: str, secret: str) -> str:
    digest = hmac.new(secret.encode("utf-8"), value.encode("utf-8"), hashlib.sha256).digest()
    return base64.urlsafe_b64encode(digest).decode("utf-8").rstrip("=")


def sign_session_token(payload: Dict[str, Any], secret: str) -> str:
    encoded_payload = base64url_encode(json.dumps(payload, separators=(",", ":")))
    signature = sign_value(encoded_payload, secret)
    return f"{encoded_payload}.{signature}"
